hyperspy/misc/model_tools.py

```diff
@@ -158,3 +160,54 @@ def _repr_html_(self):
                     html += current_component_values(
                         component=comp, only_free=self.only_free, only_active=self.only_active)._repr_html_()
         return html
+
+def linear_regression(y, component_data):
```

Comment:
Why have you taken the approach of matrix inversion rather than e.g. [`scipy.optimize.lsq_linear`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.lsq_linear.html)? Which could then (in future) support bounds?

Reply:
@tjof2 the aim of all of this is to do the full linear fitting simultaneously (it is ridiculously much faster). `lsq_linear` cannot take a multidimensional `B`. I will happily add it to support bounded fitting on a pixel by pixel basis though.

Reply:
So that is achieved by numpy array broadcasting here, right?

Reply:
In any case, what happens here when `number_of_comp > signal_axis` for the shape of `component_data`?

Reply:
Right. `numpy.linalg.lstsq` would do (although does not support bounds).

I'm thinking I will try changing to lstsq, potentially keeping both. (I have tried a lot of these in the past, but now I'm looking at this with fresh eyes)

The current approach was based on a flawed assumption that I needed to support the case where the starting value of parameters across the navigation space of a model had different values. (where `parameter.map['values']` wasn't all equal)

I've been trying to dig up an email with discussion on more motivation for this, but I think it's lost forever in my old university email account. (more evidence that all the important discussion should go in ones personal account)

Reply:
> (Assuming here you mean `Ax=b`/`Ax=B`)
yep. accidental capitalization)

Reply:
> In any case, what happens here when `number_of_comp > signal_axis` for the shape of `component_data`?

I highlight this because as currently written, `np.linalg.lstsq` and this code no longer agree on a solution, and the reasoning should be documented.

Reply:
Interesting! How did you know to try that?
I'd argue that it's a rare day where the number of components will outnumber the size of the signal axis?

I was going for speed, which is why I initially (still not completely convinced) went for matrix inversion. 
```
import numpy as np
from hyperspy.misc.model_tools import linear_regression # function in this PR

rng = np.random.RandomState(123)

signal_axis = 5000
number_of_comp = 10

A = rng.randn(signal_axis, number_of_comp)
b = rng.randn(signal_axis)
%timeit np.linalg.lstsq(A, b, rcond=None)[0]
%timeit linear_regression(b, A.T)
```
Here the latter is 8x faster on my computer, but I'm not sure how to track ram use.

Reply:
As the size of the navigation shape increases, the difference in time increases too (although so does the memory use of my function, I bet). Here's the equivalent of fitting 20 components on a <256, 256|2048> shape hyperspy signal:

```python
import numpy as np
from hyperspy.misc.model_tools import linear_regression # function in this PR

rng = np.random.RandomState(123)

signal_axis = 2048
number_of_comp = 20
nav_space = 256*256

A = rng.randn(signal_axis, number_of_comp)
b = rng.randn(nav_space, signal_axis)

AT = A.T
bT = b.T

%timeit np.linalg.lstsq(A, bT, rcond=None)[0]
%timeit linear_regression(b, AT)

np.testing.assert_array_almost_equal(
    np.linalg.lstsq(A, bT, rcond=None)[0], 
    linear_regression(b, AT).T)
```
2.94 s ± 19 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
54.2 ms ± 1.03 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

I tested with changing the size as well, and the matrix inversion function is pretty constantly 50 times faster
​

Reply:
> I highlight this because as currently written, `np.linalg.lstsq` and this code no longer agree on a solution, and the reasoning should be documented.

I completely agree on this.

Reply:
I agree that it's faster for the typical case of more data points than components. But as currently written, when data < components it gives different results to other approaches, and when data == components, it's no longer the fastest approach.

What about trying something like sklearn's `ridge_regression` - that supports multi-dimensional `y` also?

```python
import numpy as np
from scipy.linalg import lstsq as scipy_lstsq
from scipy.optimize import lsq_linear
from sklearn.linear_model._ridge import ridge_regression

rng = np.random.RandomState(123)

signal_axis = 500
number_of_comp = 500
rcond = np.finfo(float).eps * max(signal_axis, number_of_comp)

A = rng.randn(signal_axis, number_of_comp)
AT = A.T
b = rng.randn(signal_axis)

%timeit np.linalg.lstsq(A, b, rcond=rcond)
# 42.2 ms ± 2.51 ms per loop
%timeit scipy_lstsq(A, b, check_finite=False, cond=rcond, lapack_driver="gelsy")
# 8.12 ms ± 235 µs per loop
%timeit lsq_linear(A, b)
# 37.7 ms ± 668 µs per loop
%timeit ridge_regression(A, b, alpha=0.0, solver="svd")
# 48.8 ms ± 2.46 ms per loop
%timeit ridge_regression(A, b, alpha=0.0, solver="cholesky")
# 4.66 ms ± 80.9 µs per loop
%timeit ridge_regression(A, b, alpha=0.0, solver="lsqr")
# 14.4 ms ± 1.05 ms per loop
%timeit linear_regression(b, AT)
# 8.48 ms ± 70.3 µs per loop
```

Reply:
> Here's the equivalent of fitting 20 components on a `<256, 256|2048>` shape hyperspy signal

My PC is a bit slower...

```python
signal_axis = 2048
number_of_comp = 20
nav_space = 256*256

A = rng.randn(signal_axis, number_of_comp)
b = rng.randn(nav_space, signal_axis)

AT = A.T
bT = b.T

%timeit np.linalg.lstsq(A, bT, rcond=None)[0]
# 5.07 s ± 262 ms per loop
%timeit linear_regression(b, AT)
# 105 ms ± 14.3 ms per loop
%timeit ridge_regression(A, bT, alpha=0.0, solver="cholesky")
# 174 ms ± 3.09 ms per loop
%timeit ridge_regression(A, bT, alpha=0.0, solver="svd")
# 192 ms ± 6.7 ms per loop
```
Here `ridge_regression` is within 2x your inversion approach, but with a host of other potential benefits (regularization, etc.)

Reply:
Plus I just made your method fail :)

```python
import numpy as np
from scipy.sparse import rand
from sklearn.linear_model._ridge import ridge_regression

rng = np.random.RandomState(123)

signal_axis = 2048
number_of_comp = 20
nav_space = 10

A = rand(signal_axis, number_of_comp, density=1e-4, random_state=rng)
b = rng.randn(nav_space, signal_axis)

AT = A.T
bT = b.T

Ad = A.todense()
AdT = Ad.T

%timeit np.linalg.lstsq(Ad, bT, rcond=None)[0]
%timeit ridge_regression(A, bT, alpha=0.0, solver="sparse_cg")
%timeit linear_regression(b, AdT)
```
Throws `LinAlgError: Singular matrix` I'm afraid - as one would expect when using matrix inversion.

Reply:
I guess I can see how that last example could find real-life application with EDS data.

I like the sound of ridge_regularization and its benefits. I'll change to that.

Reply:
Remember that sklearn is still just an optional dependency, so it'll need a check that sklearn is installed to use it. Perhaps it could fall back to your method but with a warning of some form?

Reply:
Well remembered! Sure, that's a good idea.

Reply:
You could also add scipy.optimize.nnls into the mix as an option. Non-negativity often makes physical sense but it can result in positive bias , i.e. you always fits a small positive peak area even if there is nothing there. This becomes an issue when looking at the distribution of very dilute samples.

Reply:
@pquinn-dls yes, I will do that as a later PR (or this one, if I end up having to rework it a bit following further comments). Note that that will only allow "pixel-by-pixel" fitting, since the nnls fitter only takes one-dimensional `b`. This is in contrast to the ridge_regression and matrix inversion techniques, which with my next PR will be able to fit the whole thing at once (those take `b` of shape `(n_samples, n_targets)`.

@tjof2 I've now remembered why I originally went with the matrix inversion approach. I was doing a lot of EELS at the time. When doing DualEELS, one acquires a companion signal (low-loss) which has the same shape as the main signal. The problem here is that one has to convolve the low loss with the component array, and then fit that to the signal.

The annoying thing about that is that we go from having an `Ax = b` system where (here I'm using the [sklearn terminology](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.ridge_regression.html)) `A` has shape `(n_samples, n_features)` and b has shape `(n_samples, n_targets)` to a system where b is the same but A now has shape `(n_targets, n_samples, n_features)`. This is because every navigation pixel now is convolved by a different signal. I managed to fit that concurrently using the matrix inversion approach (using `np.matmul)`), but I haven't seen any other fitters that let me do that. 

For now (with the following PR) people will just have to do regular linear multifit (pixel-by-pixel) fitting (aka, this PR) for convoluted EELS data. I can approach the problem again at a later date.